### ISSUE TITLE
SVG output?

### ISSUE THREAD

**Issue by leon** [davidje13/lean-qr#2]
Great library with small footprint.

The only thing I'm missing is a renderer for SVG.

Any suggestions on how we could do this?

**Comment by davidje13** [comment]
You could achieve SVG output using the `.get(x,y)` function and building the SVG yourself. I wouldn't want to add this as a core feature because it would add quite a lot of code (SVGs are bulky!), but something like this should achieve what you want:

```javascript
function makeSVG(qr, scale, padding = 4) {
  const SVG_NS = 'http://www.w3.org/2000/svg';
  const svg = document.createElementNS(SVG_NS, 'svg');
  svg.setAttribute('xmlns', SVG_NS);
  svg.setAttribute('version', '1.1');
  svg.setAttribute('viewBox', `0 0 ${(qr.size + padding * 2) * scale} ${(qr.size + padding * 2) * scale}`);
  svg.setAttribute('fill', 'black');
  for (let y = 0; y < qr.size; ++y) {
    for (let x = 0; x < qr.size; ++x) {
      if (qr.get(x, y)) {
        const cell = document.createElementNS(SVG_NS, 'rect');
        cell.setAttribute('x', (x + padding) * scale);
        cell.setAttribute('y', (y + padding) * scale);
        cell.setAttribute('width', scale);
        cell.setAttribute('height', scale);
        svg.appendChild(cell);
      }
    }
  }
  return svg;
}
```

You would use this function like so:

```javascript
const qr = generate('hello');
const svg = makeSVG(qr, 10);
document.body.appendChild(svg);
```

Note that this is just a dumb implementation which draws a box for each pixel. It doesn't try to join adjacent cells into a single shape, so you will see tiny "cracks" between the cells when rendered (should still read fine but could look a little off). A quick fix might be to just make the boxes slightly bigger so they overlap, but a full fix would be lots more work.

If you want the output to be an SVG _file_ (rather than a DOM entity), you can use the same idea but building a string of `<rect x="blah" y="blah" width="blah" height="blah"></rect>`…

Hope that helps.

**Comment by leon** [comment]
Since modern bundlers such as esbuild, rollup can tree shake away functions.
why not add this as a separate function you can import that way it won't affect bundle sizes

**Comment by davidje13** [comment]
Not everybody uses a bundler; this library also works for people who just want to reference the library without a build pipeline, and since the primary goal of this project is to be a lean implementation, I'd prefer not to add any non-core stuff to the core (I think SVG export is a relatively niche requirement, albeit useful in some cases).

I decided to add the functionality under a separate import (now available in 1.3.0):

```javascript
import { generate } from 'lean-qr';
import { toSvg } from 'lean-qr/extras/svg';

const svg = toSvg(generate('hello'), document);
document.body.appendChild(svg);
```

This also implements a full algorithm for avoiding cracks between pixels. See the readme for all usage options.

Let me know if this is enough for your needs.

**Comment by davidje13** [comment]
I'm going to close this issue now since the feature has been added to the library, but feel free to reopen if it doesn't meet your needs for some reason.
